src/db_impl.cc

```diff
@@ -847,6 +847,54 @@ Status TitanDBImpl::DeleteFilesInRanges(ColumnFamilyHandle* column_family,
   return s;
 }
 
+void TitanDBImpl::MaybeScheduleRangeMerge(
+    const std::vector<std::shared_ptr<BlobFileMeta>>& files,
+    int max_sorted_runs) {
+  mutex_.AssertHeld();
+  if (files.empty()) return;
+
+  // store and sort both ends of blob files to count sorted runs
+  std::vector<std::pair<BlobFileMeta*, bool /* is smallest end? */>> blob_ends;
+  for (const auto& file : files) {
+    blob_ends.emplace_back(std::make_pair(file.get(), true));
+    blob_ends.emplace_back(std::make_pair(file.get(), false));
+  }
+  auto blob_ends_cmp = [](const std::pair<BlobFileMeta*, bool>& end1,
+                          const std::pair<BlobFileMeta*, bool>& end2) {
+    const std::string& key1 =
+        end1.second ? end1.first->smallest_key() : end1.first->largest_key();
+    const std::string& key2 =
+        end2.second ? end2.first->smallest_key() : end2.first->largest_key();
+    int cmp = key1.compare(key2);
+    // when the key being the same, order largest_key before smallest_key
+    return cmp == 0 ? end2.second : cmp < 0;
+  };
+  std::sort(blob_ends.begin(), blob_ends.end(), blob_ends_cmp);
+
+  int cur = 0;
+  int start = 0;
+  int size = blob_ends.size();
+  for (int i = 0; i < size; i++) {
+    if (blob_ends[i].second) {
+      ++cur;
+    } else {
+      // reach max sorted run in [start, i], mark these blob files to merge
+      if (cur >= max_sorted_runs) {
+        while (start < i) {
+          if (blob_ends[start].second) {
+            blob_ends[start].first->FileStateTransit(
+                BlobFileMeta::FileEvent::kNeedMerge);
+          }
+          ++start;
+        }
+      } else if (cur == 1) {
+        start = i + 1;
+      }
+      --cur;
+    }
+  }
```

Comment:
let say there are files with range [a,h], [b,c], [d,g], [e,f] and max_sorted_runs = 3, The current algorithm will mark [b,c] to merge, while it doesn't belong to a 3-sorted-run range.

Reply:
@yiwu-arbug 
Yes. I updated the algorithm to resolve this.